src/Illuminate/Database/Query/Builder.php

```diff
@@ -562,6 +562,73 @@ public function join($table, $first, $operator = null, $second = null, $type = '
         return $this;
     }
 
+    /**
+     * Join a model's table to the query while respecting its global scopes and constraints.
+     *
+     * This method enhances the standard join functionality by allowing you to join
+     * Eloquent model tables dynamically, automatically applying the model's global scopes
+     * and other query constraints. It's particularly useful for complex queries that need
+     * to maintain model-specific conditions, such as soft deletes.
+     *
+     * Key features:
+     * - Supports joining by model instance or class name string
+     * - Automatically applies all global scopes defined on the model (e.g., SoftDeletes)
+     * - Preserves model-specific query constraints
+     * - Supports all join types ('inner', 'left', 'right', etc.)
+     * - Option for 'joinWhere' functionality
+     *
+     * @param  \Illuminate\Database\Eloquent\Model|string  $model  The model instance or class name to join
+     * @param  string  $first  The first column for the join condition
+     * @param  string|null  $operator  The operator for the join condition (optional)
+     * @param  string|null  $second  The second column for the join condition (optional)
+     * @param  string  $type  The type of join ('inner', 'left', 'right', etc.)
+     * @param  bool  $where  Whether to use 'joinWhere' instead of 'join'
+     * @return $this
+     *
+     * @throws \InvalidArgumentException If the provided model is invalid
+     *
+     * @example
+     * // Join the 'users' table, respecting the User model's global scopes
+     * $query->joinModel(User::class, 'posts.user_id', '=', 'users.id');
+     */
+    public function joinModel($model, $first, $operator = null, $second = null, $type = 'inner', $where = false)
+    {
+        // If the model is a string, resolve it
+        if (is_string($model) && class_exists($model)) {
+            $model = app($model);
```

Comment:
We cannot use the `app()` helper within the Database component, as the Database component does not require the Foundation component on applications using Eloquent outside of Laravel.

You can change this line to:

```php
$model = Container::getInstance()->make($model);
```

Don't forget to import the `\Illuminate\Container\Container` class.